File: downloaders/src/main/java/com/ouattararomuald/saviezvousque/downloaders/DownloaderModule.java
package com.ouattararomuald.saviezvousque.downloaders;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
abstract class DownloaderModule {

  @Provides public static Retrofit retrofit(String baseUrl) {
    return new Retrofit.Builder()
        .baseUrl(baseUrl)
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .addConverterFactory(GsonConverterFactory.create())
        .build();
  }

  @Provides public static FeedService feedService(Retrofit retrofit) {
    return retrofit.create(FeedService.class);
  }

  @Provides public static FeedDownloader feedDownloader(FeedService feedService) {
    return FeedDownloader.Builder.create(feedService);
  }
}

File: downloaders/src/main/java/com/ouattararomuald/saviezvousque/downloaders/DownloaderModule.kt
package com.ouattararomuald.saviezvousque.downloaders

import com.ouattararomuald.saviezvousque.downloaders.FeedDownloader.Builder.create
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@InstallIn(SingletonComponent::class)
@Module
object DownloaderModule {
  @Provides
  @Singleton
  fun retrofit(): Retrofit {
    return Retrofit.Builder()
      .baseUrl(BuildConfig.FEED_BASE_URL)
      .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
      .addConverterFactory(GsonConverterFactory.create())
      .build()
  }

  @Provides fun feedService(retrofit: Retrofit): FeedService {
    return retrofit.create(FeedService::class.java)
  }

  @Provides fun feedDownloader(feedService: FeedService): FeedDownloader = create(feedService)
}